### DISCUSSION TITLE
How do you I set up the shape for Recurrent Nueral Networks?

### DISCUSSION THREAD

**Discussion by bshambaugh** [zkonduit/ezkl#164] [Q&A] [answered]
I am confused. Char-RNN[1] can have a variable length input, and that input is the training set. However, for a sufficiently long input there 
should be a fixed length vocabulary. Call this vocabulary the 128 character ASCII alphabet. 

A subset would be the input:

AAAC\n  ---> input.txt

A mapping to integers would be:

{'A':0,'C':1,'B':2,'\n':3}

Using this integer mapping translates the input to the array:
[0,0,0,2,1,3]

I see that every in https://github.com/zkonduit/ezkl/tree/main/examples/onnx is a floating point number, and not an integer. 
Also, the shape corresponds to the data length, and this is fixed. In the example above the shape would be [1,6] with the input [0.0,0.0,0.0,2.0,1.0,3.0] or
and/or [0,0,0,2,1,3] ?

If the shape corresponded to the alphabet, then the shape would be [1,128] and the input would have to be every character in the alphabet
represented once. 

[1] https://gist.githubusercontent.com/karpathy/d4dee566867f8291f086/raw/119a6930b670bced5800b6b03ec4b8cb6b8ff4ec/min-char-rnn.py

See also this repository for documentation: https://github.com/spro/char-rnn.pytorch
Ultimately I want to use the shape for the creating of an *.onnx file: https://github.com/zkonduit/pyezkl/blob/main/examples/tutorial/README.md

**Comment by bshambaugh** [comment]
On the call, Dante suggested keeping the sequence length constant.

Here is a cheat sheet for RNN.
https://stanford.edu/~shervine/teaching/cs-230/cheatsheet-recurrent-neural-networks

Here is a lecture:
http://cs231n.stanford.edu/slides/2017/cs231n_2017_lecture10.pdf

**Comment by alexander-camuto** [comment, marked as answer]
@bshambaugh check out https://github.com/zkonduit/ezkl/pull/235. This should resolve most of the issues you were facing :)

**Reply by bshambaugh** [reply]
Thanks for looking into this. I've seen some exciting Telegram news recently with stuff like nanoGPT and numerous new mathematical primitives but  I haven't dove into ezkl recently. Hopefully soon :)
